RT_SYCL/render.hpp

```diff
@@ -140,18 +95,64 @@ template <int width, int height, int samples, int depth> class render_kernel {
     }
     // If not returned within max_depth return black
     return color { 0.0f, 0.0f, 0.0f };
-  }
+  };
+
+  return [=](int x_coord, int y_coord) -> void {
+    // map the 2D indices to a single linear, 1D index
+    const auto pixel_index = y_coord * width + x_coord;
 
-  // Accessor objects
-  sycl::accessor<color, 1, sycl::access::mode::write,
-                 sycl::access::target::global_buffer>
-      m_frame_ptr;
-  sycl::accessor<hittable_t, 1, sycl::access::mode::read,
-                 sycl::access::target::global_buffer>
-      m_hitable_ptr;
-  int num_hittables;
-  camera m_camera;
-};
+    // Color sampling for antialiasing
+    color final_color(0.0f, 0.0f, 0.0f);
+    for (auto i = 0; i < samples; i++) {
+      const auto u = (x_coord + random_float()) / width;
+      const auto v = (y_coord + random_float()) / height;
+      // u and v are points on the viewport
+      ray r = cam.get_ray(u, v);
+      final_color += get_color(r);
+    }
+    final_color /= static_cast<real_t>(samples);
+
+    // Write final color to the frame buffer global memory
+    frame_ptr[pixel_index] = final_color;
+  };
+}
+
+template <int width, int height, int samples, int depth, bool single_task>
+auto get_render_kernel(sycl::accessor<color, 1, sycl::access::mode::write,
+                                      sycl::access::target::global_buffer>
+                           frame_ptr,
+                       sycl::accessor<hittable_t, 1, sycl::access::mode::read,
+                                      sycl::access::target::global_buffer>
+                           hitable_ptr,
+                       int num_hittables, camera& cam,
+                       typename std::enable_if<single_task>::type* = 0) {
+  auto pixel_kernel = pixel_renderer<width, height, samples, depth>(
+      frame_ptr, hitable_ptr, num_hittables, cam);
+  return [=]() {
+    for (int x_coord = 0; x_coord != width; ++x_coord)
+      for (int y_coord = 0; y_coord != height; ++y_coord) {
+        pixel_kernel(x_coord, y_coord);
+      }
+  };
+}
+
+template <int width, int height, int samples, int depth, bool single_task>
+auto get_render_kernel(sycl::accessor<color, 1, sycl::access::mode::write,
+                                      sycl::access::target::global_buffer>
+                           frame_ptr,
+                       sycl::accessor<hittable_t, 1, sycl::access::mode::read,
+                                      sycl::access::target::global_buffer>
+                           hitable_ptr,
+                       int num_hittables, camera& cam,
+                       typename std::enable_if<!single_task>::type* = 0) {
+  auto pixel_kernel = pixel_renderer<width, height, samples, depth>(
+      frame_ptr, hitable_ptr, num_hittables, cam);
+  return [=](sycl::nd_item<2> item) {
+    const auto x_coord = item.get_global_id(0);
+    const auto y_coord = item.get_global_id(1);
+    pixel_kernel(x_coord, y_coord);
+  };
+}
```

Comment:
This seems over-complicated.
I was thinking to `render()` just calling a
`executor(cgh, [=] (sycl::id<2> index) { /* the real render kernel is here */ })`
and having `executor` a global lambda calling either the `parallel_for` or `single_task` with the loop nest to reconstruct an `sycl::id<2>`

Reply:
Not sure if I completely understood what you wanted. 
Can you check 55ac2f9 ?